src/backend/distributed/commands/call.c

```diff
@@ -0,0 +1,170 @@
+{
+	Oid colocatedRelationId = InvalidOid;
+	Const *partitionValue = NULL;
+	ShardInterval *shardInterval = NULL;
+	List *placementList = NIL;
+	ListCell *argCell = NULL;
+	DistTableCacheEntry *distTable = NULL;
+	ShardPlacement *placement = NULL;
+	WorkerNode *workerNode = NULL;
+
+	if (IsMultiStatementTransaction())
+	{
+		ereport(DEBUG2, (errmsg("cannot push down CALL in multi-statement transaction")));
+		return false;
+	}
+
+	colocatedRelationId = ColocatedTableId(procedure->colocationId);
+	if (colocatedRelationId == InvalidOid)
+	{
+		ereport(DEBUG2, (errmsg("stored procedure does not have co-located tables")));
+		return false;
+	}
+
+	if (procedure->distributionArgIndex < 0 ||
+		procedure->distributionArgIndex >= list_length(funcExpr->args))
+	{
+		ereport(DEBUG2, (errmsg("cannot push down invalid distribution_argument_index")));
+		return false;
+	}
+
+	foreach(argCell, funcExpr->args)
+	{
+		Node *argNode = (Node *) lfirst(argCell);
+		if (!IsA(argNode, Const))
+		{
+			ereport(DEBUG2, (errmsg("cannot push down non-constant argument value")));
+			return false;
+		}
+	}
+
+	partitionValue = (Const *) list_nth(funcExpr->args, procedure->distributionArgIndex);
+	distTable = DistributedTableCacheEntry(colocatedRelationId);
+	shardInterval = FindShardInterval(partitionValue->constvalue, distTable);
+
+	if (shardInterval == NULL)
+	{
+		ereport(DEBUG2, (errmsg("cannot push down call, failed to find shard interval")));
+		return false;
+	}
+
+	placementList = FinalizedShardPlacementList(shardInterval->shardId);
+	if (list_length(placementList) != 1)
+	{
+		/* punt on reference tables for now */
+		ereport(DEBUG2, (errmsg(
+							 "cannot push down CALL for reference tables or replicated distributed tables")));
+		return false;
+	}
+
+	placement = (ShardPlacement *) linitial(placementList);
+	workerNode = FindWorkerNode(placement->nodeName, placement->nodePort);
+
+	if (workerNode == NULL || !workerNode->hasMetadata)
+	{
+		ereport(DEBUG2, (errmsg("there is no worker node with metadata")));
+		return false;
+	}
+
+	{
+		Tuplestorestate *tupleStore = tuplestore_begin_heap(true, false, work_mem);
+		TupleDesc tupleDesc = CallStmtResultDesc(callStmt);
+		TupleTableSlot *slot = MakeSingleTupleTableSlotCompat(tupleDesc,
+															  &TTSOpsMinimalTuple);
+
+		Task *task = CitusMakeNode(Task);
+		task->jobId = INVALID_JOB_ID;
+		task->taskId = 0;
+		task->taskType = DDL_TASK;
+		task->queryString = pstrdup(queryString);
+		task->replicationModel = REPLICATION_MODEL_INVALID;
+		task->dependedTaskList = NIL;
+		task->anchorShardId = placement->shardId;
+		task->relationShardList = NIL;
+		task->taskPlacementList = placementList;
+
+		ExecuteTaskListExtended(ROW_MODIFY_COMMUTATIVE, list_make1(task),
+								tupleDesc, tupleStore, true,
+								MaxAdaptiveExecutorPoolSize);
+
+		while (tuplestore_gettupleslot(tupleStore, true, false, slot))
+		{
+			if (!dest->receiveSlot(slot, dest))
+			{
+				break;
+			}
+		}
+	}
```

Comment:
Maybe we should also call `tuplestore_end(tupleStore)` here.

Reply:
I've included a comment which outlines 2 potential reasons not to call `tuplestore_end`